Add SSL support to transport layer (7.x)

apeteri opened the pull request:
The Net4j and RPC proxy implementation is replaced by Netty. When no certificate is configured in `RepositoryConfiguration`, a self-signed certificate is used &ndash; this should only be used for testing purposes.

apeteri committed:
feat(eventbus): Remove Net4j-specific packages

apeteri committed:
feat(eventbus): Update dependencies and exported packages in MANIFEST.MF

apeteri committed:
feat(eventbus): Update interfaces and utility class in public package

apeteri committed:
feat(eventbus): Update implementation classes

apeteri committed:
feat(eventbus): Add Netty channel handler implementation

apeteri committed:
feat(eventbus.tests): Update test cases

apeteri committed:
refactor(eventbus): Use replyHandler for returning results over a...

...channel; split Netty message handler into two classes; add utility
methods to EventBusNettyUtil

apeteri committed:
chore(rpc): Retire deprecated RPC project

Conflicts:
	net4j/com.b2international.snowowl.rpc.test/META-INF/MANIFEST.MF
	net4j/com.b2international.snowowl.rpc.test/pom.xml
	net4j/com.b2international.snowowl.rpc/META-INF/MANIFEST.MF
	net4j/com.b2international.snowowl.rpc/pom.xml
	net4j/com.b2international.snowowl.rpc/src/com/b2international/snowowl/internal/rpc/SecondaryInvokeRequest.java

apeteri committed:
build(netty): Rename net4j parent module to netty

Conflicts:
	net4j/com.b2international.snowowl.eventbus.tests/META-INF/MANIFEST.MF
	net4j/com.b2international.snowowl.eventbus.tests/pom.xml
	net4j/com.b2international.snowowl.eventbus/META-INF/MANIFEST.MF
	net4j/com.b2international.snowowl.eventbus/pom.xml
	net4j/pom.xml

apeteri committed:
chore(launch): Remove RPC bundle from launch configurations

Conflicts:
	fhir/com.b2international.snowowl.fhir.rest/.launch/fhir-api-server.launch

apeteri committed:
feat(core): Remove references to Net4j and RPC bundles

Conflicts:
	core/com.b2international.snowowl.core/META-INF/MANIFEST.MF
	core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/internal/locks/LockPlugin.java

apeteri committed:
feat(core): Initialize Netty server channel in RepositoryPlugin

Conflicts:
	core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/repository/RepositoryPlugin.java

apeteri committed:
feat(core): Initialize Netty client channel in TransportClient

Conflicts:
	core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/client/TransportClient.java

apeteri committed:
feat(core): Update classes related to event bus functionality

apeteri committed:
feat(netty): Use Netty's ChannelPromise for address book synchronization

apeteri committed:
fix(eventbus): Use SSL and TCP connection in test case, fix...

...synchronization issues

apeteri committed:
feat(core): Enable SSL support in RepositoryPlugin and TransportClient

Conflicts:
	core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/client/TransportClient.java
	core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/config/RepositoryConfiguration.java
	core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/repository/RepositoryPlugin.java

apeteri committed:
refactor(core): Move base event classes to eventbus

- Update subclasses and references
- Remove "send" method from BaseEvent that returns a Promise

apeteri committed:
feat(core): Remove references to deprecated RPC bundle from SnomedPlugin

apeteri committed:
feat(core): Remove class loader parameter when reading message body

apeteri committed:
feat(eventbus): Publish client connect/disconnect events

apeteri committed:
feat(eventbus): Add channel ID to messages received over a channel

apeteri committed:
feat(eventbus): Remove method IMessage#body(Class, ClassLoader)

apeteri committed:
feat(eventbus): Allow propagation of publish/send and success flags...

...in MessageFactory

apeteri committed:
refactor(core): Rename LockRequest to AbstractLockRequest; move to a...

...separate package

apeteri committed:
feat(core): Add request for lock/unlock operations using the event bus

apeteri committed:
feat(core): Restore tracking of lock targets held by a client

Most of the operations lock a branch or repository target only for a
short time interval eg. during a commit operation, but for some
authoring tasks we allow clients to issue lock requests remotely. These
locks should be released if the client disconnects and does not clean up
on its own.

apeteri committed:
fix(core): Fix issues in TransportClient

- Use NioSocketChannel instead of LocalChannel
- Extract SnowOwlConfiguration to a local variable

apeteri committed:
feat(core): Support file upload/download via event bus messaging

apeteri committed:
fix(netty): Increase Netty frame limit to 16 MiB

apeteri committed:
fix(core): Fix service registration/disposal of RemoteLockTargetListener

apeteri committed:
fix(core): Use BadRequestException when a download has already started

apeteri committed:
feat(core): Allow specifying default value in RequestHeaders

apeteri committed:
test(core): Add test cases for attachment registry client-based I/O

apeteri committed:
fix(core): Remove reference to RPC bundle from snowowl.core.feature

apeteri committed:
fix(netty): Make protocol test more robust by allowing to wait for...

...an address to appear in the address book

apeteri committed:
fix(netty): Fix formatted string issue in AddressBookNettyHandler

apeteri committed:
fix(core): Fix OutputStream issue in DefaultAttachmentRegistry

apeteri committed:
fix(lgtm): Ignore OutputStream not closed at the end of the method

Partially uploaded files are tracked in a Table and are cleaned up on
completion or "client disconnected" events.

apeteri committed:
fix(core): Fix channel handling and disposal in TransportClient

apeteri committed:
feat(netty): Disallow "half-closed" channels in EventBusNettyUtil

apeteri committed:
feat(netty): Add protocol-level keepalive messages

After 15 seconds have passed without writing anything to the channel, an
empty object is written to reset the "write idle" timer on the local
side, and the "read idle" timer on the remote. After 30 seconds of no
read activity, the corresponding channel is closed.

apeteri commented:
Re: new alert, LGTM/Semmle documentation says the following: https://lgtm.com/help/lgtm/alert-suppression

> If you've enabled automated code review of pull requests for your projects, it's important to note that alert suppression changes made in a pull request are only applied when that pull request is merged. So, if you suppress an alert in a pull request, LGTM will continue to report the related alert for as long as the pull request is open.

cmark commented on netty/com.b2international.snowowl.eventbus/src/com/b2international/snowowl/eventbus/netty/EventBusNettyUtil.java:
```diff
@@ -0,0 +1,111 @@
+				pipeline.addLast(new ObjectEncoder(), new ObjectDecoder(MAX_OBJECT_SIZE, ClassResolvers.cacheDisabled(classLoader)));
+
+				// Sends user events to handlers added later in the pipeline when there is no read/write activity
+				pipeline.addLast(new IdleStateHandler(READ_TIMEOUT_SECONDS, WRITE_TIMEOUT_SECONDS, 0));
```
It would be great to configure idle timeout values via `snowowl.yml`.

cmark reviewed (approved):
Looks great! Please check the license years and update them to 2022 if needed.

apeteri committed:
feat(netty): Make watchdog read and write timeout configurable, ...

...restore TransportConfiguration for timeouts and client-accepted
certificate property.

apeteri commented:
For the server the following sections can be configured:

```yaml
repository:
  host: 0.0.0.0
  port: 2036
  ...
  certificateChainPath: fullchain.pem     # Should include both CA and server certificate
  privateKeyPath: pkcs8.key               # Private key must be in PKCS8 format

transport:
  watchdogRate: 30       # If the connection has been idle for this long (in seconds), sends a message
  watchdogTimeout: 300   # Connection is considered broken if no data arrives for this long (seconds)
```

Client options that can be set:

```yaml
transport:
  connectionTimeout: 10       # TCP connection attempts time out after this many seconds
  watchdogRate: 30            # If the connection has been idle for this long (in seconds), sends a message
  watchdogTimeout: 300        # Connection is considered broken if no data arrives for this long (seconds)
  certificatePath: myCA.pem   # Should include the CA certificate only
```

When no certificate is given, the server uses an automatically generated, self-signed certificate; on the client, a missing `certificatePath` property results in accepting all certificates. As we don't have the identity/PKI angle figured out completely, this only covers the encryption side of the feature.

apeteri committed:
fix: Bump copyright date for changed files

apeteri committed:
fix(snomed): Register MRCM I/O classes on both server and client

apeteri committed:
fix(core): Move TransportClient registration after the channel is opened

This avoids leaking event loop groups in case multiple connection
attempts are made from the splash screen.

cmark commented:
> For the server the following sections can be configured:
> 
> ```yaml
> repository:
>   host: 0.0.0.0
>   port: 2036
>   ...
>   certificateChainPath: fullchain.pem     # Should include both CA and server certificate
>   privateKeyPath: pkcs8.key               # Private key must be in PKCS8 format
> 
> transport:
>   watchdogRate: 30       # If the connection has been idle for this long (in seconds), sends a message
>   watchdogTimeout: 300   # Connection is considered broken if no data arrives for this long (seconds)
> ```
> 
> Client options that can be set:
> 
> ```yaml
> transport:
>   connectionTimeout: 10       # TCP connection attempts time out after this many seconds
>   watchdogRate: 30            # If the connection has been idle for this long (in seconds), sends a message
>   watchdogTimeout: 300        # Connection is considered broken if no data arrives for this long (seconds)
>   certificatePath: myCA.pem   # Should include the CA certificate only
> ```
> 
> When no certificate is given, the server uses an automatically generated, self-signed certificate; on the client, a missing `certificatePath` property results in accepting all certificates. As we don't have the identity/PKI angle figured out completely, this only covers the encryption side of the feature.

Could you please add this to the docs?

apeteri commented:
The 8.x counterpart of this PR is https://github.com/b2ihealthcare/snow-owl/pull/993.
